src/NHibernate/Type/ByteType.cs

```diff
@@ -43,9 +43,10 @@ public override System.Type PrimitiveClass
 
 		public override void Set(DbCommand cmd, object value, int index, ISessionImplementor session)
 		{
-			cmd.Parameters[index].Value = Convert.ToByte(value);
-		}
+			var dp = cmd.Parameters[index];
+			dp.Value = dp.DbType == DbType.Int16 ? Convert.ToInt16(value) : Convert.ToByte(value);
```

Comment:
DB2 does not support byte native (as many RDBMS) and has a problem handling it. If the value is of type byte the provider remaps the parameter as Binary. Then it throws `InvalidCastException` trying to cast the value to `byte[]`.